### PR TITLE
#21: Add Kokkos dependency for magistrate

### LINKED ISSUE #21: add optional Kokkos support to magistrate
`distBVH` uses spack packages in CI and it fails when [attempting to serialize `Kokkos::View`](https://github.com/sandialabs/distBVH/actions/runs/10187718320/job/28182385646?pr=26):
```
/opt/view/include/checkpoint/dispatch/dispatch_serializer_nonbyte.h:233:16: error: no matching function for call to 'checkpoint::dispatch::SerializerDispatchNonByte<checkpoint::Sizer, Kokkos::View<unsigned char*, Kokkos::LayoutRight, Kokkos::OpenMP>, checkpoint::dispatch::BasicDispatcher<checkpoint::Sizer, Kokkos::View<unsigned char*, Kokkos::LayoutRight, Kokkos::OpenMP> > >::applyStatic(checkpoint::Sizer&, Kokkos::View<unsigned char*, Kokkos::LayoutRight, Kokkos::OpenMP>*&, checkpoint::SerialSizeType&)'
```

**TODO**:
Add `Kokkos` as an optional dependency and corresponding variant (`default=False`) to `magistrate` package.

`magistrate` enables Kokkos support based on existence of `Kokkos_DIR` or `Kokkos_ROOT`, so I'm assuming something like `"-DKokkos_ROOT={}".format(self.spec["kokkos"].prefix)` will be necessary.

### PR DESCRIPTION
**Pull request by JacobDomagala**
Fixes #21

### CODE REVIEW

**Comment by JacobDomagala** [comment]
@cz4rs @nmm0 Let me know if that works for you. I don't use Spack so not sure whether you want to build everything from scratch or you'd rather use existing Kokkos installation.

File: packages/darma-magistrate/package.py [hunk 1]
```diff
@@ -43,8 +43,21 @@ class DarmaMagistrate(CMakePackage):
...
-    sanity_check_is_file = ["cmake/magistrateConfig.cmake"]
-    sanity_check_is_file = ["cmake/magistrateTargets.cmake"]
-    sanity_check_is_file = ["lib/libmagistrate.a"]
+    sanity_check_is_file = [
+        "cmake/magistrateConfig.cmake",
+        "cmake/magistrateTargets.cmake",
+        "lib/libmagistrate.a"
+    ]
+
+    depends_on("kokkos", when="+kokkos_enabled")
```
Comment (cz4rs) [inline packages/darma-magistrate/package.py:55]: This results in a pretty old version of Kokkos:
```
==> Installing kokkos-3.2.00-kse4d7zqrzcnajlg675fo6gryrnknkky
```

I think we should pin to something from 4.x series. `4.3.1` was released in May and for what it's worth CI in `distBVH` uses `4.1`.

**Review by cz4rs — APPROVED** [review]
This looks pretty reasonable to me. Minimal experience with Spack here though :wink:

File: packages/darma-magistrate/package.py (hunk 1 above)
Comment (JacobDomagala) [reply packages/darma-magistrate/package.py]: `spack info kokkos` gives me the following wrt versioning:
```
Preferred version:  
    3.2.00     https://github.com/kokkos/kokkos/archive/3.2.00.tar.gz

Safe versions:  
    develop    [git] https://github.com/kokkos/kokkos.git on branch develop
    master     [git] https://github.com/kokkos/kokkos.git on branch master
    3.2.00     https://github.com/kokkos/kokkos/archive/3.2.00.tar.gz
    3.1.01     https://github.com/kokkos/kokkos/archive/3.1.01.tar.gz
    3.1.00     https://github.com/kokkos/kokkos/archive/3.1.00.tar.gz
    3.0.00     https://github.com/kokkos/kokkos/archive/3.0.00.tar.gz    
```

Comment (JacobDomagala) [reply packages/darma-magistrate/package.py]: `master` branch fails with:
```
    'cmake' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:STRING=/__w/spack-package/spack-package/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/kokkos-master-ndrkjrvtrnur4ltnv6tqzyqsagazw64d' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION:BOOL=OFF' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=OFF' '-DCMAKE_INSTALL_RPATH:STRING=/__w/spack-package/spack-package/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/kokkos-master-ndrkjrvtrnur4ltnv6tqzyqsagazw64d/lib;/__w/spack-package/spack-package/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-10.5.0/kokkos-master-ndrkjrvtrnur4ltnv6tqzyqsagazw64d/lib64' '-DCMAKE_PREFIX_PATH:STRING=/cmake' '-DKokkos_ARCH_ZEN2=ON' '-DKokkos_ENABLE_CUDA=OFF' '-DKokkos_ENABLE_OPENMP=OFF' '-DKokkos_ENABLE_PTHREAD=OFF' '-DKokkos_ENABLE_SERIAL=ON' '-DKokkos_ENABLE_HIP=OFF' '-DKokkos_ENABLE_AGGRESSIVE_VECTORIZATION=OFF' '-DKokkos_ENABLE_COMPILER_WARNINGS=OFF' '-DKokkos_ENABLE_CUDA_LAMBDA=OFF' '-DKokkos_ENABLE_CUDA_LDG_INTRINSIC=OFF' '-DKokkos_ENABLE_CUDA_RELOCATABLE_DEVICE_CODE=OFF' '-DKokkos_ENABLE_CUDA_UVM=OFF' '-DKokkos_ENABLE_DEBUG=OFF' '-DKokkos_ENABLE_DEBUG_BOUNDS_CHECK=OFF' '-DKokkos_ENABLE_DEBUG_DUALVIEW_MODIFY_CHECK=OFF' '-DKokkos_ENABLE_DEPRECATED_CODE=OFF' '-DKokkos_ENABLE_EXAMPLES=OFF' '-DKokkos_ENABLE_EXPLICIT_INSTANTIATION=OFF' '-DKokkos_ENABLE_HPX_ASYNC_DISPATCH=OFF' '-DKokkos_ENABLE_PROFILING=ON' '-DKokkos_ENABLE_PROFILING_LOAD_PRINT=OFF' '-DKokkos_ENABLE_QTHREAD=OFF' '-DKokkos_ENABLE_TESTS=OFF' '-DKokkos_ENABLE_HPX=OFF' '-DKokkos_ENABLE_HWLOC=OFF' '-DKokkos_ENABLE_NUMACTL=OFF' '-DKokkos_ENABLE_MEMKIND=OFF' '-DCMAKE_CXX_COMPILER=/__w/spack-package/spack-package/spack/lib/spack/env/gcc/g++' '-DKokkos_CXX_STANDARD=11' '-DBUILD_SHARED_LIBS=True' '/tmp/root/spack-stage/spack-stage-kokkos-master-ndrkjrvtrnur4ltnv6tqzyqsagazw64d/spack-src'

1 error found in build log:
  >> 3    CMake Error at cmake/kokkos_pick_cxx_std.cmake:12 (MESSAGE):
     4      Setting the variable Kokkos_CXX_STANDARD in configuration is deprec
          ated -
     5      set CMAKE_CXX_STANDARD directly instead
     6    Call Stack (most recent call first):
     7      CMakeLists.txt:48 (INCLUDE)
     8    
     9    
```

Comment (JacobDomagala) [reply packages/darma-magistrate/package.py]: Looking at https://github.com/kokkos/kokkos/blob/develop/Spack.md I think the build error could be somehow fixed with proper variants in `packages.yaml` (which is local file)

Comment (JacobDomagala) [reply packages/darma-magistrate/package.py]: OK nevermind the previous messages, we were using older version of Spack...

**Comment by JacobDomagala** [comment]
Ok finally got it working. Currently it uses latest release (v4.3.01), should we pin it to 4.1.0?

**Comment by cz4rs** [comment]
> Ok finally got it working. Currently it uses latest release (v4.3.01), should we pin it to 4.1.0?

I think latest is fine, users can specify the version they want anyways. I was only surprised with the `3.2.0`.

**Review by lifflander — APPROVED** [review]
Looks good to me

**Review by nmm0 — APPROVED** [review]
Looks good to me.

Only a minor comment is we could rename `kokkos_enabled` to just `kokkos` since that's in line with other spack packages

**Pull request merged**